NETOBSERV-526 Display milliseconds for time fields in NetFlow table

jpinsonneau opened the pull request:
This PR adds miliseconds in both table / side panel.

In table, date are using short formats like `10/11/2022 12:56:28.208
` fitting into a single row.
Long format is used in side panel `Tue Oct 11, 2022 12:56:28.208 PM
`

jotak commented on web/src/components/netflow-record/record-field.tsx:
```diff
@@ -122,8 +124,12 @@ export const RecordField: React.FC<{
+      return emptyText();
+    }
+
+    const dateText = getFormattedDate(date, longDates ? 'ddd ll' : 'MM/DD/YYYY');
```
that's the american format? Can we have it based on browser locale? (if not painful)

stleerh replied:
If you don't mind using moment.js, then do something like this so it considers locale and handles milliseconds.  Example:  
  
`const shortFormat = moment('2022-10-11 08:25:01.123').format('l hh:mm:ss.SSS');`  
`const longFormat = moment('2022-10-11 08:25:01.123').format('ddd ll h:mm:ss.SSS A');`  
  
`moment` accepts many formats for the date/time argument or leave it empty if you want the time now.

jpinsonneau replied:
Moved to Intl.DateTimeFormat

stleerh commented on web/src/components/netflow-record/record-field.tsx:
```diff
@@ -122,8 +124,12 @@ export const RecordField: React.FC<{
+      return emptyText();
+    }
+
+    const dateText = getFormattedDate(date, longDates ? 'ddd ll' : 'MM/DD/YYYY');
```
Code still has 'MM/DD/YYYY' and should be 'l'.

stleerh commented on web/src/components/netflow-table/__tests__/netflow-table.spec.tsx:
```diff
@@ -70,7 +71,7 @@ describe('<NetflowTable />', () => {
     // table should be sorted by date asc by default
     expect(wrapper.find(NetflowTableRow).find(Td).at(timestampIdx).find('.datetime').at(0).text()).toBe(
-      expectedDate.toDateString() + ' ' + expectedDate.toLocaleTimeString()
+      getFormattedDate(expectedDate, 'MM/DD/YYYY HH:mm:ss.SSS')
```
Should be:  
getFormattedDate(expectedDate, 'l HH:mm:ss.SSS')

stleerh commented on web/src/components/netflow-table/__tests__/netflow-table.spec.tsx:
```diff
@@ -80,7 +81,7 @@ describe('<NetflowTable />', () => {
     // then should sort date desc on click
     expect(wrapper.find(NetflowTableRow).find(Td).at(timestampIdx).find('.datetime').at(0).text()).toBe(
-      expectedDate.toDateString() + ' ' + expectedDate.toLocaleTimeString()
+      getFormattedDate(expectedDate, 'MM/DD/YYYY HH:mm:ss.SSS')
```
Should be:  
getFormattedDate(expectedDate, 'l HH:mm:ss.SSS')

Amoghrd commented:
/ok-to-test

stleerh commented on web/src/utils/datetime.ts:
```diff
@@ -42,6 +42,9 @@ export const getTimeRangeOptions = (t: TFunction) => {
 export const getFormattedDate = (date: Date, format = 'llll') => {
+  if (window.navigator.language.length && moment.locale() !== window.navigator.language) {
+    moment.locale(window.navigator.language);
+  }
```
What's the reason for doing this?  On my machine, moment.locale() is "en" and window.navigator.language is "en-US".  Therefore, it will call moment.locale('en-US'), but that ends up getting changed to "en", and so this will get set on every call.

Also, you can drop "window." and just use navigator.language since it's implied.

jpinsonneau replied:
Moved to `navigator.language` + Intl.DateTimeFormat

andrew-ronaldson commented:
I was talking with @jpinsonneau about this yesterday but adding comments here. We can add the milliseconds but the date display format should match the rest of console. We would need a pretty solid reason to break from that console convention.

stleerh commented:
Sorry, we should not use moment.js after all since it's in [maintenance mode](https://momentjs.com/docs/#/-project-status/), and there are more modern alternatives.  One such alternative is [Intl.DateTimeFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat), which is supported by all modern browsers.

stleerh commented:
> I was talking with @jpinsonneau about this yesterday but adding comments here. We can add the milliseconds but the date display format should match the rest of console. We would need a pretty solid reason to break from that console convention.

Just to be clear, this is the date/time format I see in most of the Console panels but with seconds and milliseconds added.  Example:  
Oct 13, 2022, 9:20:15.413 AM  
  
With this _date format_, which is pretty much we have now, it avoids localization issues.  The changes from what we have in Network Observability are:  
\- Remove day such as "Thu".  
\- Add a comma after the month day (e.g. 13) and after the year.  
\- Add milliseconds.

jpinsonneau commented:
> Sorry, we should not use moment.js after all since it's in [maintenance mode](https://momentjs.com/docs/#/-project-status/), and there are more modern alternatives. 

Using moment is still fine as they are supporting it:
```
However, since we understand that Moment is well established in millions of existing projects:
- We will address critical security concerns as they arise.
- We will release data updates for Moment-Timezone following [IANA time zone database](https://www.iana.org/time-zones) releases. 
```

However, for consistency with console, I moved to `Intl.DateTimeFormat`. This needed to update typescript and configure `esnext` capabilities.

As you requested:
- Remove day such as "Thu".
- Add a comma after the month day (e.g. 13) and after the year.
- Add milliseconds.

Also added from console:
- Icons
- Tooltips with UTC format

Small size + tooltip
![image](https://user-images.githubusercontent.com/91894519/195815237-b62504f5-5dd9-4777-9a70-10671f5242d6.png)

Normal / large size with not enough room for single line
![image](https://user-images.githubusercontent.com/91894519/195815514-c48cd18d-8a2d-4011-8fb3-d20dfa0ca679.png)

French format
![image](https://user-images.githubusercontent.com/91894519/195815798-2d3c1b48-b168-4aa0-98c6-23c48d11cc0d.png)

Amoghrd commented:
/ok-to-test

Amoghrd commented:
/qe-approved

jpinsonneau commented on web/src/utils/datetime.ts:
```diff
@@ -41,8 +41,60 @@ export const getTimeRangeOptions = (t: TFunction) => {
+  day: 'numeric',
+  year: 'numeric'
+});
+
```
For me yes. In the table we don't show the weekday and in the side panel we shows it.

Do you have something else in mind ?

jpinsonneau replied:
I removed it for consistency https://github.com/netobserv/network-observability-console-plugin/pull/205/commits/b9570c43d86959992994cd09d68002e1e63d9bc7

![image](https://user-images.githubusercontent.com/91894519/196649591-d061026c-10a4-4877-9f87-8ec25407a2b5.png)

jpinsonneau replied:
thanks @stleerh !

Amoghrd commented:
/ok-to-test

Amoghrd commented:
/qe-approved

jpinsonneau commented:
/approve

stleerh commented:
/lgtm
